validator/core/src/output/data_type.rs

```diff
@@ -0,0 +1,356 @@
+    Compound(Compound),
+
+    /// User-defined type.
+    UserDefined(Arc<extension::Reference<extension::DataType>>),
+
+    /// Unresolved type. Used for error recovery.
+    Unresolved(String),
+}
+
+impl std::fmt::Display for Class {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Class::Simple(simple) => write!(f, "{}", simple),
+            Class::Compound(compound) => write!(f, "{}", compound),
+            Class::UserDefined(user_defined) => write!(f, "{}", user_defined),
+            Class::Unresolved(name) => write!(f, "{}!", name),
+        }
+    }
+}
+
+impl ParameterChecker for Class {
+    fn check_parameters(&self, params: &[Parameter]) -> diagnostic::Result<()> {
+        match self {
+            Class::Simple(_) => {
+                if params.is_empty() {
+                    Ok(())
+                } else {
+                    Err(cause!(
+                        TypeMismatchedParameters,
+                        "simple types cannot be parameterized"
+                    ))
+                }
+            }
+            Class::Compound(compound) => compound.check_parameters(params),
+            Class::UserDefined(_) => {
+                if params.is_empty() {
+                    Ok(())
+                } else {
+                    Err(cause!(
+                        TypeMismatchedParameters,
+                        "user-defined types cannot currently be parameterized"
+                    ))
+                }
+            }
+            Class::Unresolved(_) => Ok(()),
+        }
+    }
+}
+
+/// Enumeration of simple types defined by Substrait.
+#[derive(Clone, Debug, PartialEq, Display, EnumString)]
+#[strum(ascii_case_insensitive, serialize_all = "snake_case")]
+pub enum Simple {
+    Boolean,
+    I8,
+    I16,
+    I32,
+    I64,
+    Fp32,
+    Fp64,
+    String,
+    Binary,
+    Timestamp,
+    TimestampTz,
+    Date,
+    Time,
+    IntervalYear,
+    IntervalDay,
+    Uuid,
+}
+
+/// Enumeration of compound types defined by Substrait.
+#[derive(Clone, Debug, PartialEq, Display, EnumString)]
+#[strum(ascii_case_insensitive, serialize_all = "UPPERCASE")]
+pub enum Compound {
+    FixedChar,
+    VarChar,
+    FixedBinary,
+    Decimal,
+    Struct,
+    #[strum(serialize = "NSTRUCT")]
+    NamedStruct,
+    List,
+    Map,
+}
+
+impl ParameterChecker for Compound {
+    fn check_parameters(&self, params: &[Parameter]) -> diagnostic::Result<()> {
+        match self {
+            Compound::FixedChar | Compound::VarChar | Compound::FixedBinary => {
+                if params.len() != 1 {
+                    return Err(cause!(
+                        TypeMismatchedParameters,
+                        "{} expects a single parameter (length)",
+                        self
+                    ));
+                }
+                if let Parameter::Unsigned(length) = params[0] {
+                    if length < 1 && length > 2147483647 {
```

Comment:
```suggestion
                    if length < 1 && length > i32::MAX {
```

Reply:
I think I did it this way because the spec explicitly lists this as the range. It might as well have listed 10000000 or whatever.

Reply:
We can put one of them in a comment? One matches the spec, the other makes it easier to understand where that number comes from.

Reply:
Added comment and used a local const for it to deduplicate.